### ISSUE TITLE
Feature - support third party (non Amazon) S3 for snapshot agent backups 

### ISSUE THREAD

**Issue by pannon** [hashicorp/consul#3707]
We are Consul enterprise users and would welcome a feature enhancement to the snapshot agent.
More specifically, support third party (non Amazon) S3 storage target for the snapshot agent. 

There is a growing number of S3 compatible Opensource alternatives.
The requested feature should essentially allow snapshot saving to any S3 compatible storage (LeoFS, Minio, Riak CS, etc).
 
The storage we use is [LeoFS](https://github.com/leo-project/leofs)
Some details on their [client example in Go](https://leo-project.net/leofs/docs-old/s3api_client/s3_client_golang.html)

**Comment by slackpad** [comment]
Hi @pannon thanks for opening an issue. Have you tried the current snapshot agent with LeoFS and can you describe any errors you get? I'm wondering if there's some simple tweaks we can make to use the right subset of the S3 API that these support or if we'd need to pull in a new client library.

**Comment by pannon** [comment]
Hi @slackpad, thanks for picking this up. I have attempted to configure the current snapshot agent with LeoFS and hit two issues:
1. missing option for custom endpoint configuration
2. option to use self signed certs (disable SSL verification/or disable SSL completely)

For 1. I've attempted to "fake" Amazon's (endpoint) host record, this resulted in a TLS/SSL error. If SSL verification could be disabled it could potentially work (this is not the best solution).

Best would be to add a feature for custom endpoints (and ports) with an option to turn SSL verification off or disable SSL completely. 

There are some details [here on configuring s3cmd with LeoFS](https://leo-project.net/leofs/docs-old/s3api_client/s3_client_s3cmd.html), where the actual LeoFS endpoint is substituted for the HTTP proxy option and SSL is turned off. Maybe adding an HTTP proxy to the snapshot agent is also an option.  

In terms of S3 API compatibility, the best is to assume basic functionality with these 3rd party solutions. For example S3 ACLs, snapshots or other sophisticated features might not be supported at all.
